climada/util/yearsets.py

```diff
@@ -22,223 +22,227 @@
+        events_per_year = sample_n_events(n_sampled_years, imp, lam)
+        sampling_vect = sample_events(events_per_year, imp.frequency)
 
+    #compute impact per sampled_year
+    imp_per_year = compute_imp_per_year(imp, sampling_vect)
 
-    #create sampling dictionary if not given as input
-    if not sampling_dict:
-        n_annual_events = np.sum(event_impacts.frequency)
-        n_input_events = len(event_impacts.event_id)
-        sampling_dict = create_sampling_dict(n_sampled_years, n_annual_events,
-                                             n_input_events)
+    #copy imp object as basis for the yimp object
+    yimp = copy.deepcopy(imp)
 
-    #compute annual_impacts
-    impact_per_year = compute_annual_impacts(event_impacts, sampling_dict)
-
-    #copy event_impacts object as basis for the annual_impacts object
-    annual_impacts = copy.deepcopy(event_impacts)
-
-    #save impact_per_year in annual_impacts
+    #save imp_per_year in yimp
     if correction_fac: #adjust for sampling error
-        correction_factor = calculate_correction_fac(impact_per_year, event_impacts)
-        annual_impacts.at_event = impact_per_year / correction_factor
+        correction_factor = calculate_correction_fac(imp_per_year, imp)
+        yimp.at_event = imp_per_year / correction_factor
     else:
-        annual_impacts.at_event = impact_per_year
+        yimp.at_event = imp_per_year
 
-    annual_impacts.event_id = np.arange(1, n_sampled_years+1)
-    annual_impacts.tag['annual_impacts object'] = True
-    annual_impacts.date = u_dt.str_to_date([str(date) + '-01-01' for date in sampled_years])
-    annual_impacts.frequency = np.ones(n_sampled_years)*np.sum(sampling_dict['events_per_year']
+    #save calculations in yimp
+    yimp.event_id = np.arange(1, n_sampled_years+1)
+    yimp.tag['yimp object'] = True
+    yimp.date = u_dt.str_to_date([str(date) + '-01-01' for date in sampled_years])
+    yimp.frequency = np.ones(n_sampled_years)*sum(len(row) for row in sampling_vect
                                                             )/n_sampled_years
 
+    return yimp, sampling_vect
 
-    return annual_impacts, sampling_dict
-
-def create_sampling_dict(n_sampled_years, n_annual_events, n_input_events):
-    """Create a sampling dictionary consisting of the amount of events per sample year and the
-    index of the sampled events
+def sample_n_events(n_sampled_years, imp, lam=None):
+    """Create a sampling vector consisting of a subarray for each sampled_year that
+    contains the index of the sampled events for that year
 
     Parameters:
         n_sampled_years : int
             The target number of years the impact yearset shall contain.
-        n_annual_events : int
-            Number of events per year in given event_impacts object
-        n_input_events : int
-            Number of events contained in given event_impacts object
+        imp : climada.engine.Impact()
+            impact object containing impacts per event
```

Comment:
I think this is not needed in this function. `lam = np.sum(imp.frequency)` should be used in the function that calls this one. You could then also rename the method as `sample_from_poisson` or similar.